### PR TITLE
fix: preserve file URL targets in Action Fusion

### LINKED ISSUE #2: Action Fusion skips then_run after a successful write to a file URL
## Summary

Action Fusion independently resolves the target path used for its post-mutation hash check. Its resolver does not handle `file://` URLs, although the wrapped Pi 0.84.2 `write` and `edit` tools do. As a result, a fused mutation using a valid file URL writes the requested file successfully but then skips `then_run` with `ENOENT`.

## Reproduction

On commit `8f8c13916c97fb54c32008e41c05421f9b69ac3b`, Node.js `v22.22.2`, npm `10.9.7`:

```js
import { mkdtemp, readFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { pathToFileURL } from "node:url";
import { createActionFusionExtension } from "./src/sol-pi/extensions/action-fusion/index.ts";

const tools = new Map();
createActionFusionExtension({
  bashOptions: { operations: { exec: async () => ({ exitCode: 0 }) } },
})({ registerTool: (tool) => tools.set(tool.name, tool) });

const cwd = await mkdtemp(join(tmpdir(), "solpi-file-url-"));
const target = join(cwd, "target.txt");
const context = {
  cwd,
  mode: "json",
  hasUI: false,
  model: undefined,
  sessionManager: {
    getSessionId: () => "repro",
    getSessionFile: () => undefined,
  },
  ui: {},
};

try {
  await tools.get("write").execute(
    "call",
    {
      path: pathToFileURL(target).href,
      content: "written\n",
      then_run: { command: "true" },
    },
    undefined,
    undefined,
    context,
  );
} catch (error) {
  console.log(String(error));
}
console.log(await readFile(target, "utf8"));
```

Run from the repository root with:

```bash
node --experimental-strip-types repro.mjs
```

## Actual result

The mutation succeeds and `target.txt` contains `written\n`, but the tool call reports:

```text
Error: [then_run:skipped] ENOENT: no such file or directory, open '.../file:/.../target.txt'; the command was not run.
```

`src/sol-pi/extensions/action-fusion/file-queue.ts:16-20` passes the URL string to `node:path.resolve()`, producing a path containing `file:`. That path is later read by `assertUnchangedBeforeCommand()`.

Pi 0.84.2's wrapped mutation tools use `resolveToCwd()`, whose underlying path normalization converts `file://` inputs with `fileURLToPath()`. Therefore the mutation and Action Fusion's guard operate on different paths.

## Expected result

Action Fusion should apply the same path semantics as the Pi mutation tool it wraps, so the hash check targets the file that was actually written and `then_run` executes.

This likely also warrants parity tests for the other normalization handled by Pi's resolver, such as Unicode spaces and Windows shell-style drive paths, but only the `file://` case above was reproduced on this host.

### PR DESCRIPTION
**Pull request by Owen718**
## Summary

Fixes #2.

Decode `file://` targets with Node's `fileURLToPath()` before resolving Action Fusion's file queue and post-mutation hash guard. Pi already accepts these URLs; the guard must inspect the same file that Pi wrote or edited.

This keeps the existing queue, hash checks and command-failure behavior intact. It does not patch Pi, import private Pi APIs, or change dependencies.

## Regression coverage

Ten new tests cover file URLs and `@file://` inputs, percent-encoded spaces/`#`/`%`, real writes and edits, a real bash follow-up, serialization between URL and ordinary-path aliases, invalid URL rejection, and unchanged ordinary-path handling.

Before the runtime fix, five of the initial seven regression tests failed, including successful file mutations followed by `then_run:skipped`/`ENOENT`. The corrected implementation passes the expanded suite.

## Validation

- `npm ci --ignore-scripts`: passed; lockfile unchanged.
- `npm run check`: passed, 132 tests across 17 files, plus typecheck and package dry run.
- `node scripts/check-pi-compat.mjs`: passed.
- `npx --no-install vitest run tests/all-mechanisms.test.ts`: passed, 4 tests.
- `npm audit --audit-level=high`: passed. The unchanged lockfile currently has two moderate Vitest-related advisories; no dependency changes are included.

Tested with Node 26.0.0 and Pi 0.84.2. Based directly on `main`, independent of the renderer-test fix for #1; validation used a checkout path without the case-sensitive `SoL-Pi` substring.

### CODE REVIEW

File: src/sol-pi/extensions/action-fusion/file-queue.ts [hunk 1]
```diff
@@ -14,7 +15,9 @@ function stripToolPathPrefix(filePath: string): string {
 }
 
 export function resolveToolPath(cwd: string, filePath: string): string {
-	const expanded = stripToolPathPrefix(filePath);
+	const stripped = stripToolPathPrefix(filePath);
+	// Pi accepts file URLs; the queue and hash guard must use the same target.
+	const expanded = stripped.startsWith("file://") ? fileURLToPath(stripped) : stripped;
```
Comment (QihangCao) [inline src/sol-pi/extensions/action-fusion/file-queue.ts:20]: **[P2] Malformed `file://` URLs lose the `[then_run:skipped]` marker**

`resolveToolPath()` now calls `fileURLToPath(stripped)` here, but it's evaluated as an argument to `executeMutationThenRun(...)` in `index.ts:93` / `index.ts:127` — i.e. **before** that function's try/catch even starts. If the URL is malformed (e.g. the `%2F`-encoded case your own new test covers), `fileURLToPath` throws synchronously and the error bypasses both skip-wrapping sites (`then-run.ts:97-104` around `mutate()`, and `assertUnchangedBeforeCommand`'s own catch at `then-run.ts:54-68`).

Before this PR, the same malformed input would have failed inside `mutate()` (Pi's own resolver rejecting it) and gotten properly wrapped with `[then_run:skipped]`. That marker is a tested contract elsewhere in this repo (`tests/action-fusion.test.ts:289`, `:414`), so this is a real, if narrow, regression against it — just not covered by the new test, which only asserts `.rejects.toThrow()` without checking the message.

Confirmed reproducible:
```
fileURLToPath("file:///.../bad%2Fname.txt")
→ TypeError: File URL path must not include encoded / characters
```

Suggested fix: move the `fileURLToPath` decode inside a guarded path that still produces `[then_run:skipped]` on failure — e.g. catch it in `resolveToolPath`'s caller and translate to the skipped-result shape, or do the decode lazily inside `executeMutationThenRun`'s existing try/catch instead of at the call site.

No data-loss or security impact — the existing test already confirms no mutation/command runs. This is purely an error-reporting/UX regression for one edge case.

**Review by QihangCao — COMMENTED** [review]
Core fix is correct — it resolves #2 (file URLs no longer wrongly skip `then_run`). Found one edge-case regression in error reporting for malformed file URLs; not blocking, but worth a follow-up before merge.
